File: backend/api/routes/resumes.py
from fastapi import APIRouter, HTTPException

from db.resume_repository import (
    create_resume,
    delete_resume,
    list_resumes,
    get_resume,
)
from schemas.resume import (
    ResumeCreateRequest,
    ResumeCreateResponse,
    ResumeDeleteResponse,
    ResumeRecord,
    ResumeSummary,
)

router = APIRouter(prefix="/resumes", tags=["resumes"])


@router.get("", response_model=list[ResumeSummary])
async def get_all_resumes():
    return await list_resumes()


@router.get("/{resume_id}", response_model=ResumeRecord)
async def get_resume_by_id(resume_id: str):
    try:
        resume = await get_resume(resume_id)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")

    return resume


@router.post("", response_model=ResumeCreateResponse)
async def create_resume_endpoint(body: ResumeCreateRequest):
    resume_id = await create_resume(body.name, body.resume_data.model_dump())

    return ResumeCreateResponse(id=resume_id)


@router.delete("/{resume_id}", response_model=ResumeDeleteResponse)
async def delete_resume_by_id(resume_id: str):
    try:
        deleted = await delete_resume(resume_id)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    if not deleted:
        raise HTTPException(status_code=404, detail="Resume not found")

    return ResumeDeleteResponse(deleted=True)

File: frontend/src/lib/api/resumes.ts
import type { ResumeData } from "@/lib/types";

export type ResumeSummary = {
  id: string;
  name: string;
  created_at?: string;
  updated_at?: string;
};

export type ResumeRecord = {
  id: string;
  name: string;
  resume_data: ResumeData;
  created_at?: string;
  updated_at?: string;
};

type ResumeCreateResponse = {
  id: string;
};

// Same-origin proxy avoids CORS in dev; override to hit FastAPI directly if needed.
const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? "/api";
const DIRECT_API_BASE =
  process.env.NEXT_PUBLIC_DIRECT_API_BASE_URL ?? "http://127.0.0.1:8000";
const DEV_FALLBACK_ENABLED =
  process.env.NODE_ENV !== "production" && API_BASE.startsWith("/");
const RESUMES_URL =
  process.env.NEXT_PUBLIC_RESUMES_URL ?? `${API_BASE}/resumes`;
const DIRECT_RESUMES_URL = `${DIRECT_API_BASE}/resumes`;
const RESUMES_PRIMARY_URL = DEV_FALLBACK_ENABLED
  ? DIRECT_RESUMES_URL
  : RESUMES_URL;
const RESUMES_FALLBACK_URL = DEV_FALLBACK_ENABLED
  ? RESUMES_URL
  : DIRECT_RESUMES_URL;

async function parseErrorDetail(
  response: Response,
  fallback: string,
): Promise<string> {
  try {
    const body = (await response.json()) as {
      detail?: string | { msg?: string }[];
    };
    if (typeof body.detail === "string") return body.detail;
    if (Array.isArray(body.detail) && body.detail[0]?.msg) {
      return body.detail.map((d) => d.msg).join("; ");
    }
  } catch {
    // ignore JSON parse errors
  }
  return fallback;
}

async function fetchWithDevFallback(
  primaryUrl: string,
  fallbackUrl: string,
  init: RequestInit,
): Promise<Response> {
  if (!DEV_FALLBACK_ENABLED || primaryUrl === fallbackUrl) {
    return fetch(primaryUrl, init);
  }

  try {
    const response = await fetch(primaryUrl, init);
    if (response.ok || response.status < 500) {
      return response;
    }
  } catch {
    // Ignore fetch errors and try the direct backend in dev.
  }

  return fetch(fallbackUrl, init);
}

export async function saveResume(
  name: string,
  resumeData: ResumeData,
): Promise<ResumeCreateResponse> {
  const response = await fetchWithDevFallback(
    RESUMES_PRIMARY_URL,
    RESUMES_FALLBACK_URL,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name, resume_data: resumeData }),
    },
  );

  if (!response.ok) {
    const detail = await parseErrorDetail(
      response,
      `Save failed (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ResumeCreateResponse;
}

export async function listResumes(): Promise<ResumeSummary[]> {
  const response = await fetchWithDevFallback(
    RESUMES_PRIMARY_URL,
    RESUMES_FALLBACK_URL,
    { method: "GET" },
  );

  if (!response.ok) {
    const detail = await parseErrorDetail(
      response,
      `Load resumes failed (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ResumeSummary[];
}

export async function getResume(resumeId: string): Promise<ResumeRecord> {
  const response = await fetchWithDevFallback(
    `${RESUMES_PRIMARY_URL}/${resumeId}`,
    `${RESUMES_FALLBACK_URL}/${resumeId}`,
    { method: "GET" },
  );

  if (!response.ok) {
    const detail = await parseErrorDetail(
      response,
      `Load resume failed (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ResumeRecord;
}

export async function deleteResume(resumeId: string): Promise<void> {
  const response = await fetchWithDevFallback(
    `${RESUMES_PRIMARY_URL}/${resumeId}`,
    `${RESUMES_FALLBACK_URL}/${resumeId}`,
    { method: "DELETE" },
  );

  if (!response.ok) {
    const detail = await parseErrorDetail(
      response,
      `Delete failed (${response.status})`,
    );
    throw new Error(detail);
  }
}
